pkg/pillar/types/zedroutertypes.go

```diff
@@ -2889,3 +2922,102 @@ func (status OnboardingStatus) LogDelete(logBase *base.LogObject) {
 func (status OnboardingStatus) LogKey() string {
 	return string(base.OnboardingStatusLogType) + "-" + status.Key()
 }
+
+// Bitmap :
+// Bitmap of the reserved and allocated resources
+// Keeps 256 bits indexed by 0 to 255.
+type Bitmap [32]byte
+
+// IsSet :
+// Test the bit value
+func (bits *Bitmap) IsSet(i int) bool {
+	return bits[i/8]&(1<<uint(7-i%8)) != 0
+}
+
+// Set :
+// Set the bit value
+func (bits *Bitmap) Set(i int) {
+	bits[i/8] |= 1 << uint(7-i%8)
+}
+
+// Clear :
+// Clear the bit value
+func (bits *Bitmap) Clear(i int) {
+	bits[i/8] &^= 1 << uint(7-i%8)
+}
+
+// AddToIP :
+func AddToIP(ip net.IP, addition int) net.IP {
+	if addr := ip.To4(); addr != nil {
+		val := uint32(addr[0])<<24 + uint32(addr[1])<<16 +
+			uint32(addr[2])<<8 + uint32(addr[3])
+		val += uint32(addition)
+		byte0 := byte((val >> 24) & 0xFF)
+		byte1 := byte((val >> 16) & 0xFF)
+		byte2 := byte((val >> 8) & 0xFF)
+		byte3 := byte(val & 0xFF)
+		return net.IPv4(byte0, byte1, byte2, byte3)
+	}
+	//TBD:XXX, IPv6 handling
+	return net.IP{}
+}
+
+// GetIPSubnetBase :
+func GetIPSubnetBase(ip, gateway net.IP) int {
+	if addr := ip.To4(); addr != nil {
+		val0 := uint32(addr[0])<<24 + uint32(addr[1])<<16 +
+			uint32(addr[2])<<8 + uint32(addr[3])
+		if gateway == nil {
+			return 0
+		}
+		addr = gateway.To4()
+		if addr == nil {
+			return 0
+		}
+		val1 := uint32(addr[0])<<24 + uint32(addr[1])<<16 +
+			uint32(addr[2])<<8 + uint32(addr[3])
+		return int(val1 - val0)
+	}
+	//TBD:XXX, IPv6 handling
+	return 0
+}
+
+// GetIPAddrCountOnSubnet IP address count on subnet
+func GetIPAddrCountOnSubnet(subnet net.IPNet) int {
+	prefixLen, _ := subnet.Mask.Size()
+	if prefixLen != 0 {
+		if subnet.IP.To4() != nil {
+			return 0x01 << (32 - prefixLen)
+		}
+		if subnet.IP.To16() != nil {
+			return 0x01 << (128 - prefixLen)
+		}
+	}
+	return 0
+}
+
+// GetIPNetwork  :
+func GetIPNetwork(subnet net.IPNet) net.IP {
+	return subnet.IP.Mask(subnet.Mask)
+}
+
+// GetIPBroadcast :
+func GetIPBroadcast(subnet net.IPNet) net.IP {
+	if network := GetIPNetwork(subnet); network != nil {
+		if addrCount := GetIPAddrCountOnSubnet(subnet); addrCount != 0 {
+			return AddToIP(network, addrCount-1)
+		}
+	}
+	return net.IP{}
+}
+
+// AppNumber :
+// PS. Any change to AppNumMax, must be
```

Comment:
I think AppNumMax should be replaced by a BitMapMax since this is really a constraint due to the bitmap.
And that max should be 255 since that is the largest which fits in in the bitmap.

Note that the -3 is already handled in the parseIpspec to ensure that the subnet, broadcast, and gateway are excluded from dhcprange.